File: bootcamps/bootcamp_2020/solutions/exercise1_solution.py
############################################################################
############################################################################
#	exercise1_solution.py
#
#	September 2020,  baclark@msu.edu
#	Example solution for the first exercise in bootcamp 2020
#
#	just to mix it up, and demonstrate a more python-oriented analysis flow
#	unlike in the C++ version, I am not going to rely on any ROOT classes 
#	after I have accessed the waveforms
# 	I will also use matplotlib to make the plot
############################################################################

import ROOT
import numpy as np
import matplotlib.pyplot as plot
import os

ROOT.gSystem.Load(os.environ.get('ARA_UTIL_INSTALL_DIR')+"/lib/libAraEvent.so")

# open a data file
file = ROOT.TFile.Open("/data/wipac/ARA/2014/unblinded/L1/ARA02/0101/run002823/event002823.root")

# load in the event free for this file
eventTree = file.Get("eventTree")

# set the tree address to access our raw data type
rawEvent = ROOT.RawAtriStationEvent()
eventTree.SetBranchAddress("event",ROOT.AddressOf(rawEvent))

# get the number of entries in this file
num_events = eventTree.GetEntries()
print('total events:', num_events)

list_of_snrs = []

# loop over the events
for event in range(100):
	
	# get the event
	eventTree.GetEntry(event)

	#if not a cal pulser, continue
	if(rawEvent.isCalpulserEvent()==0): 
		continue

	# make a useful event
	usefulEvent = ROOT.UsefulAtriStationEvent(rawEvent,ROOT.AraCalType.kLatestCalib)

	graph = usefulEvent.getGraphFromElecChan(0)

	# unpack the voltages into a numpy array to make it easier to do math on
	# I'm going to do this to model the case where we want to stop using ROOT
	# after we've gotten the raw waveform

	volts = []
	for sample in range(graph.GetN()):
		volts.append(graph.GetY()[sample])
	volts = np.asarray(volts)

	# now, I can use np to compute the rms and the peak voltage

	rms = np.std(volts)
	peak = np.amax(volts)

	list_of_snrs.append(peak/rms)

	del usefulEvent

list_of_snrs = np.asarray(list_of_snrs)

# now we make a histogram
fig = plt.figure(figsize=(10,5))
ax = fig.add_subplot(111)
ax.hist(list_of_snrs, bins=20)
ax.set_ylabel('SNR')
ax.set_xlabel('Number of Events')
ax.set_title('Histogram of Ch 0 SNRs')
fig.savefig('hist_of_snr_py.png', bbox_inches='tight', dpi=300)

File: bootcamps/bootcamp_2020/solutions/exercise1_solution.cpp
////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
////	exercise1_solution.cpp
////
////	September 2020,  baclark@msu.edu
////	Example solution for the first exercise in bootcamp 2020
////
////	In this example, I will use ROOT to make a histogram
////	And also rely on class member functions of the TGraph to help
////	calculate the SNRs
////////////////////////////////////////////////////////////////////////////////

// C/C++ Includes
#include <iostream>

//AraRoot Includes
#include "RawAtriStationEvent.h"
#include "UsefulAtriStationEvent.h"

//ROOT Includes
#include "TTree.h"
#include "TFile.h"
#include "TGraph.h"
#include "TCanvas.h"
#include "TH1D.h"
#include "TMath.h"

using namespace std;

int main(int argc, char **argv)
{

	// for output, we will make a histogram
	TH1D *h = new TH1D("snrs","snrs",20,0,10);

	if(argc<2) {  // Check to make sure there are enough arguments to do something meaningful
		std::cout << "Usage requires you to provide input parameter of the form " << basename(argv[0]) << " <input data file>" << std::endl;
		return -1;
	}
	
	// open a data file, where the data file was passed as the first argument to the executable
	TFile *fpIn = new TFile(argv[1], "OLD");
	if(!fpIn){
		std::cerr<< "Can not open the old file: " <<argv[1]<<endl;
		return -1;
	} //throw a warning if you can't open it
	
	fpIn->cd(); //go into that file
	TTree *eventTree = (TTree*) fpIn->Get("eventTree"); //load in the event free for this file
	if(!eventTree){
		std::cerr<<"Can't find eventTree in file" <<argv[1]<<" with filename " <<argv[1]<<endl;
		return -1;
	} //throw a warning if you can't open it
	
	//set the tree address to access our raw data type
	RawAtriStationEvent *rawEvent=0;
	eventTree->SetBranchAddress("event",&rawEvent);
	
	double numEntries = eventTree -> GetEntries(); //get the number of entries in this file

	for(int event=0; event<numEntries; event++){ //loop over those entries
		
		eventTree->GetEntry(event); //get the event

		bool isCalpulser = rawEvent->isCalpulserEvent();

		if(!isCalpulser) continue;
		
		//make a *useful* event out of the *raw* event, which functionally just calibrates it
		UsefulAtriStationEvent *usefulEvent = new UsefulAtriStationEvent(rawEvent, AraCalType::kLatestCalib);
	
		//now, we'll get the waveform from channel 0
		TGraph *waveform = usefulEvent->getGraphFromRFChan(0);

		// because it's a TGraph, we can use some TMath utilities to calculate
		// the peak and rms, and that allows to calculate the SNR

		double rms = TMath::RMS(waveform->GetN(), waveform->GetY());

		// now, before computing the peak, ge absolute values 
		for(int sample=0; sample<waveform->GetN(); sample++) waveform->GetY()[sample] = abs(waveform->GetY()[sample]);

		double peak = TMath::MaxElement(waveform->GetN(), waveform->GetY());	

		h->Fill(peak/rms);

		delete waveform;
		delete usefulEvent;
	}

	TCanvas *c = new TCanvas("c","c", 1100,850);
	h->Draw("hist");
	h->SetTitle("Histogram of Ch 0 SNRs; SNR; Number of Events");
	c->SaveAs("hist_of_snr_cpp.png");

}//close the main program
